gnocchi/rest/aggregates/operations.py

```diff
@@ -32,39 +36,202 @@
 }
 
 
-def handle_aggregate(agg, granularity, timestamps, values, is_aggregated):
+# TODO(sileht): expose all operators in capability API
+binary_operators = {
+    u"=": numpy.equal,
+    u"==": numpy.equal,
+    u"eq": numpy.equal,
+
+    u"<": numpy.less,
+    u"lt": numpy.less,
+
+    u">": numpy.greater,
+    u"gt": numpy.greater,
+
+    u"<=": numpy.less_equal,
+    u"≤": numpy.less_equal,
+    u"le": numpy.less_equal,
+
+    u">=": numpy.greater_equal,
+    u"≥": numpy.greater_equal,
+    u"ge": numpy.greater_equal,
+
+    u"!=": numpy.not_equal,
+    u"≠": numpy.not_equal,
+    u"ne": numpy.not_equal,
+
+    u"%": numpy.mod,
+    u"mod": numpy.mod,
+
+    u"+": numpy.add,
+    u"add": numpy.add,
+
+    u"-": numpy.subtract,
+    u"sub": numpy.subtract,
+
+    u"*": numpy.multiply,
+    u"×": numpy.multiply,
+    u"mul": numpy.multiply,
+
+    u"/": numpy.true_divide,
+    u"÷": numpy.true_divide,
+    u"div": numpy.true_divide,
+
+    u"**": numpy.power,
+    u"^": numpy.power,
+    u"pow": numpy.power,
+
+}
+
+# TODO(sileht): adds, numpy.around, but it take a decimal argument to handle
+unary_operators = {
+    u"abs": numpy.absolute,
+    u"absolute": numpy.absolute,
+
+    u"neg": numpy.negative,
+    u"negative": numpy.negative,
+
+    u"cos": numpy.cos,
+    u"sin": numpy.sin,
+    u"tan": numpy.tan,
+    u"floor": numpy.floor,
+    u"ceil": numpy.ceil,
+}
+
+
+def handle_unary_operator(nodes, granularity, timestamps, initial_values,
+                          is_aggregated, references):
+    op = nodes[0]
+    granularity, timestamps, values, is_aggregated = evaluate(
+        nodes[1], granularity, timestamps, initial_values,
+        is_aggregated, references)
+
+    values = unary_operators[op](values)
+    return granularity, timestamps, values, is_aggregated
+
+
+def handle_binary_operator(nodes, granularity, timestamps,
+                           initial_values, is_aggregated, references):
+    op = nodes[0]
+    g1, t1, v1, is_a1 = evaluate(nodes[1], granularity, timestamps,
+                                 initial_values, is_aggregated, references)
+    g2, t2, v2, is_a2 = evaluate(nodes[2], granularity, timestamps,
+                                 initial_values, is_aggregated, references)
+
+    is_aggregated = is_a1 or is_a2
+    # We keep the computed timeseries
+    if isinstance(v1, numpy.ndarray) and isinstance(v2, numpy.ndarray):
+        if not numpy.array_equal(t1, t2) or g1 != g2:
+            raise exceptions.UnAggregableTimeseries(
+                references,
+                "Can't compute timeseries with different "
+                "granularity %s <> %s" % (nodes[1], nodes[2]))
+        timestamps = t1
+        granularity = g1
+        is_aggregated = True
+
+    elif isinstance(v1, numpy.ndarray):
+        timestamps = t1
+        granularity = g1
+    elif isinstance(v1, numpy.ndarray):
```

Comment:
isinstance(v2, numpy.ndarray)?

seems like we're missing a test here.

Reply:
Triggering that bug was not so easy, and was not raising a 500, but producing wrong set of data with the first row correct ... on the test I was expecting to take that elif branch. Then processor.get_measures uses zip(), and silently dropped unexpected data!

So I have added a sanity check to evaluate() to check the returning data and raise RuntimeError if timestamps/values/granularity doesn't match anymore.